ichnaea/service/locate.py

```diff
@@ -122,496 +100,536 @@ def map_data(data):
+    LOCATION_PROVIDER_CLASSES = (
+        CellAreaLocationProvider,
+        CellLocationProvider,
+        WifiLocationProvider,
+    )
+
+    def __init__(self, session, geoip_db, *args, **kwargs):
+        super(AbstractLocationSearcher, self).__init__(*args, **kwargs)
+        self.session = session
+        self.geoip_provider = GeoIPLocationProvider(
+            api_key_log=self.api_key_log,
+            api_key_name=self.api_key_name,
+            api_name=self.api_name,
+            geoip_db=geoip_db,
+            session=self.session,
+        )
+        self.heka_client = get_heka_client()
+
+        self.location_providers = [
+            cls(
+                api_key_log=self.api_key_log,
+                api_key_name=self.api_key_name,
+                api_name=self.api_name,
+                session=self.session
+            ) for cls in self.LOCATION_PROVIDER_CLASSES]
+
+    def search_location(self, data, client_addr):
+        location = None
+        location_provider_used = None
+
+        # Always do a GeoIP lookup because it is cheap and we want to
+        # report geoip vs. other data mismatches. We may also use
+        # the full GeoIP City-level estimate as well, if all else fails.
+        geoip_location, country = self.geoip_provider.locate(client_addr)
+
+        for location_provider in self.location_providers:
+            provider_location = location_provider.locate(data)
+
+            if provider_location:
+                if location is None:
+                    # If this is our first hit, then we use it.
+                    location = provider_location
+                    location_provider_used = location_provider
+                else:
+                    # If this location is more accurate than our previous one,
+                    # we'll use it.
+                    provider_distance = distance(
+                        location['lat'],
+                        location['lon'],
+                        provider_location['lat'],
+                        provider_location['lon']
+                    ) * 1000
+                    if provider_distance <= location['accuracy']:
+                        location = provider_location
+                        location_provider_used = location_provider
+
+        # Fall back to GeoIP if nothing has worked yet. We do not
+        # include this in the "zoom-in" loop because GeoIP is
+        # frequently _wrong_ at the city level; we only want to
+        # accept that estimate if we got nothing better from cell
+        # or wifi.
+        if not location and geoip_location:
+            location = geoip_location
+            location_provider_used = self.geoip_provider
+
+        if not location:
+            self.stat_count('miss')
+
+        for location_provider in\
+                self.location_providers + [self.geoip_provider]:
+            if location_provider is location_provider_used:
+                location_provider.log_used()
```

Comment:
Thinking about this, maybe each provider can return an additional three-state value, for example: True == provider was used and hit, False == provider was used and missed, None == provider wasn't used at all. The first provider which returns a `is not None` result can then update the location_provider_used variable to (class, True/False) value. The following providers won't update the location_provider_used variable anymore, as it's not None anymore.